pkg/netlink/eventbus/eventbus.go

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022-2023 Intel Corporation, or its subsidiaries.
+
+// Package eventbus handles pub sub
+package eventbus
+
+import (
+	"sync"
+)
+
+// EventBus holds the event bus info
+type EventBus struct {
+	subscribers map[string][]*Subscriber
+	mutex       sync.RWMutex
+}
+
+// Subscriber holds the info for each subscriber
+type Subscriber struct {
+	Ch   chan interface{}
+	Quit chan bool
+}
+
+// NewEventBus initializes ann EventBus object
+func NewEventBus() *EventBus {
+	return &EventBus{
+		subscribers: make(map[string][]*Subscriber),
+	}
+}
+
+// Subscribe api provides registration of a subscriber to the given eventType
+func (e *EventBus) Subscribe(eventType string) *Subscriber {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+
+	subscriber := &Subscriber{
+		Ch:   make(chan interface{}),
+		Quit: make(chan bool),
+	}
+
+	e.subscribers[eventType] = append(e.subscribers[eventType], subscriber)
+
+	return subscriber
+}
+
+// Publish api notifies the subscribers with certain eventType
+func (e *EventBus) Publish(eventType string, data interface{}) {
+	e.mutex.RLock()
+	defer e.mutex.RUnlock()
+
+	subscribers, ok := e.subscribers[eventType]
+	if !ok {
+		return
+	}
+
+	for _, sub := range subscribers {
```

Comment:
I still do not understand why it is not possible.
Do we have a sequence diagram demonstrating the flow?

Reply:
replaced Rlock to Lock in publish function as we dont have multiple reader adnd writer to same channel, also modified unsubscriber function for deleting the map entries So that Netlink do not publish to the channel if its closed

Reply:
if you don't use locks for reads in RWMutex, then you should change the type to Mutex
and again, if no one is listening on sub.Ch? We will block and wait until someone starts listening there, which may never happen(a client is trying to call Unsubscribe, event bus is trying to send data to client(which is not listening))
Should we use [non-blocking channel operations](https://gobyexample.com/non-blocking-channel-operations) trying to send data here?

Reply:
Changed to Mutex.

Reply:
but what about?
>It is bad to block holding a mutex. It can result in a deadlock.
Should we use [non-blocking channel operations](https://gobyexample.com/non-blocking-channel-operations) trying to send data here?

When we just -> to chan, it will block. So we block holding the mutex. Let's use non-blocking chanoperations

Reply:
We are aware of it and are actively working on a solution. To maintain the momentum of our current development cycle, we propose not to consider this issue a blocker for the current PR. We plan to address it in a subsequent PR. 
 issue link : https://github.com/opiproject/opi-evpn-bridge/issues/388

Reply:
Ack